sdk/python/pvsite_datamodel/write/generation.py

```diff
@@ -1,97 +1,73 @@
-"""Functions for writing to pvsite db."""
-import datetime as dt
+"""
+Write helpers for the Generation table.
+"""
+
 import logging
-import uuid
 
-import numpy.typing as npt
 import pandas as pd
-import pandera as pa
-import sqlalchemy.orm as sa_orm
+import sqlalchemy as sa
+from sqlalchemy.dialects import postgresql
+from sqlalchemy.orm import Session
 
-from pvsite_datamodel.read.site import get_site_by_uuid
-from pvsite_datamodel.sqlmodels import GenerationSQL
+from pvsite_datamodel.sqlmodels import Base, GenerationSQL
 
-# Defines the length of time over which a forecast is valid
-from pvsite_datamodel.write.datetime_intervals import get_or_else_create_datetime_interval
-from pvsite_datamodel.write.upsert import upsert
-from pvsite_datamodel.write.utils import UUIDV4, WrittenRow
+_log = logging.getLogger(__name__)
 
 
-class GenerationValuesSchema(pa.SchemaModel):
-    """Schema for the dataframe used by the insert_generation_values function."""
+def _upsert(session: Session, table: Base, rows: list[dict]):
+    """Upserts rows into table.
 
-    start_utc: pa.typing.Series[pd.DatetimeTZDtype] = pa.Field(
-        dtype_kwargs={"unit": "ns", "tz": "UTC"},
-        description="The start of the time period over which the generation value is valid."
-    )
-    power_kw: pa.typing.Series[pa.dtypes.Float] = pa.Field(
-        ge=0,
-        description="The power generated by the site in kw over the given time period."
-    )
-    site_uuid: pa.typing.Series[UUIDV4] = pa.Field(
-        description="The UUID of the site to which the generation value refers."
+    This functions checks the primary keys, and if present, updates the row.
+    :param session: sqlalchemy Session
+    :param table: the table
+    :param rows: the rows we are going to update
+    """
+    stmt = postgresql.insert(table.__table__)
+    primary_key_names = [key.name for key in sa.inspect(table.__table__).primary_key]
+    update_dict = {c.name: c for c in stmt.excluded if not c.primary_key}
+
+    if not update_dict:
+        raise ValueError("insert_or_update resulted in an empty update_dict")
+
+    stmt = stmt.on_conflict_do_update(
+        index_elements=primary_key_names, set_=update_dict
     )
+    session.execute(stmt, rows)
 
 
 def insert_generation_values(
```

Comment:
How come we've lost the schema for the dataframe here?

Reply:
Yea, I think simon wanted to keep it simon, I personally like the schema checker, so maybe it could be 2:1 on this. @simlmx what do you think?

Pros and cons of keeping schema in
**Pros**
- good validation,
- good validation should lead to clear error messages

**Cons**:
- more code, slight more complicated
- another dependencies in the repo

Reply:
First I don't think this code ("write" helper functions) should live in this repo at all: it should be closer to the code using it. I kept it in here for now because it was a tricky refactor. This is for another discussion but I think that is we want different services to interact with the DB in a well defined uniform way, they should probably do it through the API, not through a common helper function. And if two services really need to both interact with the database in a similar fashion, they should probably implement their own helper functions, even if it means having copy-pasted functions at first: those could very well diverge in the future.

We shouldn't need the same kind of validation that we use when receiving data from an external source (like a client sending data to an API) here, for the simple reason that we control the code that creates the data. **We don't need to handle the errors from bad data** (unlike for client-facing APIs), **we simply want to crash because if the data is wrong, it means our own code is broken**. This already happens without any validation: if the type of the data or the column names are wrong, the `INSERT/UPDATE` will fail.

I would also argue that errors like `IndexError: 'column X' not in dataframe` or `DBDataError: column X is expecting a Datetime, Float provided` will be clear enough in this context.

I no one is convinced by my arguments I'm happy to bring the pydantic validation back 😛

Reply:
Perhaps just put int some asserts for the moment, and we can come back to it another time

Reply:
Yes actually if some *development* errors are easy to make and would not be caught by some other SqlalchemyError/IndexError/etc., then some `assert`s for sanity checks would make a lot of sense.

I think here we are covered by "normal" errors though, or am I forgetting some possible error types?